### DISCUSSION TITLE
Does .net maui have brighter future?

### DISCUSSION THREAD

**Discussion by manishbisen** [dotnet/maui#10839] [Q&A]
I am planning to explore my skills in mobile cross platform development So my question is will it be worth to start  exploring with .net maui.Does it have bright future?Honestly speaking  this question comes in my mind because when I check updates in my twitter or Youtube ,I see many developers negative feedback releated to bugs in .net maui & some of positive feedback also.So this reasons have make me confused to start from where.Or should I go with other framework like flutter or react native ?

**Comment by nicjay** [comment]
See the following for previous community discussion: 
https://github.com/dotnet/maui/discussions/4879
https://github.com/dotnet/maui/discussions/9478
https://github.com/dotnet/maui/discussions/10596 (Notably, this has an excellent answer by David Ortinau)

Personal opinion:

There are bugs. There are bugs that make me wonder why they're even still there in core features. They also dropped the ball on tooling (VS for Mac is STILL only preview MAUI support). If you need to build a stable product in the next few months, I would look elsewhere.

However, it's .NET, and that's a very, very strong technology to work on these days. Both C# and .NET push the envelope with major yearly releases. I have no doubt about the technical foundations of MAUI underneath the surface bugs, and I think post .NET 7 and into 2023 we'll see it really start to take off.

**Comment by holecekp** [comment]
My personal opinion is that MAUI is not worth learning in its current state. Even the most basic controls are full of bugs. The performance is also terrible compared to the old Xamarin.Forms. If you start learning MAUI now, it would only discourage you because you would spend the most of the time by searching Github Issues instead of the actual learning.

If you are already a .NET developer, I would maybe consider learning Xamarin.Forms instead. Yes, it is legacy now. But it is a really great framework, it is easy to learn, and you can reuse your existing .NET code and skills. Moreover, it is almost the same as MAUI with only a few differences. So it should not be very difficult to migrate from Xamarin.Forms to MAUI later when it is ready. You would need to learn just a few more things (e.g., handlers) but the basics are exactly the same.

I really hope that MAUI will get better in the upcoming year.

**Comment by mphill** [comment]
Microsoft uses React Native extensively.  They  basically never use Xamarin/MAUI for their apps.

Teams, Outlook, Office - all leverage React Native.

https://reactnative.dev/showcase

The planning and resource allocation seems utterly bizarre to me. I would love for someone at MS to comment on this.

**Reply by BirukTes** [reply]
> Microsoft uses React Native extensively. They basically never use Xamarin/MAUI for their apps.

I guess, Microsoft must demonstrate with their own use for their produce, until we can move on to any of these wagons and deem them with having a bright future. Makes you wonder, why even touch it when they are not?

**Reply by akhanalcs** [reply]
Exactly!

**Comment by ToolmakerSteve** [comment]
**tl;dr: .Net has a bright future. Whether Maui does is not yet knowable. If you want to learn a successful .Net based platform today, use Unity3D. Great preparation for VR/AR. Otherwise, learn Flutter. That will set you on a solid path, both web and elsewhere.**

I love .Net CLR. At that core level, it has huge potential for making programmers more productive. I especially love the way that .Net supports C#, F# (functional + OO), and dynamic languages.

When it was open sourced, I was ecstatic. This was the paradigm I wanted future software to be developed on. When Xamarin "Native" brought great performance on mobile platforms, this became even stronger. Can develop Android and iOS apps effectively, programming only in c#.

However, Xamarin "Forms" (which I have used heavily for years) never became "good enough". I estimate that overall, it has NOT been a net gain in programming productivity, compared to developing the UI twice using Xamarin "Native". Far too many problems and limitations to work around.

IMHO, it is not yet knowable whether Maui will reach "good enough". It hasn't yet; in addition to hundreds of issues/limitations still there from XF, there are new issues popping up every month or two. Go to StackOverflow, and enter tag `[maui]`. You'll see. You'll also see some dedicated people, both inside and outside Microsoft, volunteering their time to help solve the problems, and to say when to "add this as a new issue at github maui issues".

My point is, its a work in process. I dearly wish Microsoft was putting in MUCH more development resources to Maui than they are. OTOH, there is sufficient passion around this, both in and out of Microsoft, that it WILL pull through.

Now, for the bigger problem. The web. The web is by far the most important platform. If you care about cross-platform, then you should include Web as one of your primary targets. Today, to me, that means using either Flutter or (Typescript + ReactJS).

True, there is Blazor. I would only use Blazor if I was working in an Enterprise environment, with a heavy investment in Microsoft-based tools and solutions. I don't consider it in the same category as Flutter or ReactJS; it only makes sense (IMHO) if you are using ASP.Net as your server solution. That isn't a terrible career path, but it isn't the path I would suggest someone start on today.

Most importantly, compare Flutter as a cross-platform solution, to Blazor as a cross-platform solution. Flutter is well-architected, and has been very favorably received, is rapidly growing in popularity. IMHO, Blazor by itself simply doesn't cut it. And Maui Blazor is better than nothing, but isn't a joy to work with. 

The only way .Net makes sense as a long-term winner, is if it works well on the web, and has a cross-platform GUI. "Someday" Maui will reach that status. But notice the lack of web support for Maui. This is beyond unfathomable, from a business perspective.

BTW, there are THREE .Net-based UIs that work today on web: Unity 3D, AvaloniaUI, and UnoPlatform.

Unity 3D is a massively successful company due to games. Which is a perfect starting point for VR/AR.  
If I was starting today, I'd start there.

I don't know why Microsoft didn't pour money into one of the open source solutions (AvaloniaUI or UnoPlatform), so .Net would have a cross-platform (including web) UI today.

**Comment by coderbasket** [comment]
I think .Net Maui is not ready for some... look at this ,,, my code have no errors but the app refuse to compile....
![Screenshot 2022-11-26 223339](https://user-images.githubusercontent.com/97447955/204094118-0f3e49b0-bcdf-4fd7-a596-f2256d711141.png)

**Comment by coderbasket** [comment]
![Screenshot 2022-11-30 103950](https://user-images.githubusercontent.com/97447955/204694608-8e3e4397-4e45-498f-a579-9c0a8d7cb979.png)

**Comment by nextcodelab** [comment]
I have not done any project in .Net Maui... all my projects going back to zero because of some issues.
I will try Avalonia I will reply here my experience...

**Reply by liebki** [reply]
Who said you need to use MAUI?

**Reply by nextcodelab** [reply]
I really need .Net because all my projects libraries written on it...  and I am very comportable with C#... And still waiting for some improvement of .Net Maui... After maybe some testing in coming months I will decide if what UI frameworks in .Net  families to use.
That's why I created this simple sites for all .net projects. 
https://coderbasket.github.io/

**Comment by SF-Simon** [comment]
I tried it for a while, and then looked forward to it for a long time. I hope MAUI has new hope, but every time I pay attention to it, I am disappointed.

Why is it still using XAML? XAML feels like it's back in the 19th century. Where is Blazor's progress?

Maybe it can't be regarded as a toy, because toys has fun, but MAUI doesn't.

This may be the last time I paid attention to it. Even if C# is good, I have no patience to watch MAUI. goodbye.

**Reply by 98Beka** [reply]
u can use blazor without problem in MAUI project

**Reply by mphill** [reply]
https://github.com/dotnet/maui/issues/12124

**Comment by Mistofs** [comment]
No. Use anything else until MAUI is 100% working. There's so many alternative why get yourself wet when you can use the vast amount of javascript libraries to do the same stuff.

**Reply by Mistofs** [reply]
You can use C# you can use F# while still developing cross platform UI without touching MAUI
MAUI.blazor is a bit better, but why not just blazor, why not just compile F# to javascript

**Comment by nextcodelab** [comment]
My solution to all of these,, is rewriting the app from maui to flutter...
Flutter super fast almost 1 second in startup... any designs is posible;
https://play.google.com/store/apps/details?id=com.nxtccode.biblenotebook

**Reply by Hooterr** [reply]
I kinda got used to XAML (the core idea and work flow is fine, it's just the framework that's broken) and I like C# but I guess they do everything for me to start hating it over time...

**Reply by liebki** [reply]
But Flutter is not C#, most companies can't just change the language and framework that easily.

**Comment by npostma** [comment]
I've developed several Xamarin apps that have been published over the last 8 years. In my opinion, it was never the best framework, but it got the job done, and it provided a consistent experience across different platforms. Occasionally, you needed to make minor platform-specific adjustments. I chose Xamarin because I'm familiar with .NET/C#, so it was the logical choice. However, MAUI is relatively new and still under heavy development. Currently, I'm struggling to port some of my apps to MAUI without much success. I expected that at least the basic functionality of Xamarin apps would be seamlessly integrated into MAUI. But that is not yet the case for now. You have to see it as if its a new product. For now, it is what it is, and you have to use workarounds (and sometimes workarounds on them) to make things work. It seems that they've picked up the pace with .NET 8, as I've noticed many updates being released. So, I'm now trying to contribute to MAUI wherever I can to expedite my progress, and I'm keeping my fingers crossed for a smoother transition.

**Comment by Hooterr** [comment]
That is so true. Xamarin was not great, but it worked for the most part. You had to spend some time to really polish it but you could get there at the end.
But this... They didn't even manage to copy paste existing functionality without creating a whole bunch of new issues. Tooling constantly breaks. The workloads not having the option to choose a version of MAUI is a total fucking disaster. Like, how the hell did you come up with this colossally stupid idea?

**Reply by janseris** [reply]
You can choose MAUI version using a secret XML tag in .csproj: `<MauiVersion></MauiVersion>`

**Reply by Hooterr** [reply]
That should've been there from day 1. I don't know how you can release something and not realize such a fatal flaw in your design.

**Comment by AlleSchonWeg** [comment]
I'm still on xf. And i'm afraid to port my app to maui. We implemented a lot of workarounds in xf and i'm not sure if they fixed in maui or not.  We also use some third party libs like https://github.com/CrossGeeks/FirebasePushNotificationPlugin, which are incompatible. Also the iOS google xf bindings looks abandoned (https://github.com/xamarin/GoogleApisForiOSComponents) and i read nothing if iOS notification service extensions will work in maui.

**Comment by admoss0** [comment]
[I wrote a blog entry on this today ](https://admoss.blogspot.com/2023/07/is-net-maui-waste-of-time.html). When you finally try to deploy a .NET Maui application in the real world it will come as a shock to realise it can't be done without an AUD $550 per year code signing certificate. Over $500 a year to run your own software on your own computer? Hard pass.

**Reply by vpenades** [reply]
I am not sure if that certificate is related to MAUI (or .Net) at all, We've been selling .Net app for years not using maui at all, and we do need to pay for that certificate.

You can use a free self-certificate, the difference between a self-certificate and a paid one is that if, for example, you're creating a windows installer, on paid certificate the installer will run straight away, but on a self-certificate, a yellow warning window saying "windows cannot verify this installer, proceed at your own risk" window before the installer can proceed.  This is similar on app stores.

**Reply by admoss0** [reply]
It is specific to Maui. I have a number of .Net applicstions in use At work without the need for code signing.

You cannot just bundle up the publish folder of. a Maiu applicstion and distribute it. It won’t run.

 As to self signed certificates, they sre not trusted. You hsve to manually load them into a trusted partition using the certificate snapin in mmc. Most corporate users cannot do this as they lack the rights. Many home users lack the skills to do it.

 Besides, putting a self signed cert from a random stranger into a trusted psrtition is not good security practice.

So you need a code signing certificate from a trusted source. That costs AUD  $550 per year.

You cannot create an installer for a Maui app without a certificate in Visual Studio. It isnt the exe we know and love. It is an MSIX file.

**Reply by liebki** [reply]
I created a few MAUI applications for me and GitHub, they all run without certificate. You obviously didn't try well enough.

You can run a MAUI program on your local hardware without paying or buying a certificate..

Posted an article about how to:

https://medium.com/@kxmliebl/how-to-publish-a-maui-application-locally-on-your-device-79c0fbd12f7e

**Reply by Xyncgas** [reply]
The thing the guy is getting at is there is currently no option for deploying to binary executable on the target platform as a stand alone app. You have to publish it to app stores in every platform to get the best experience or you can let people install it as an 'app' that's supposed to be more modern than the good old '.exe' but it requries you to have the certificate for this modern experience.

Can we just not have this modern experience though

**Reply by Hooterr** [reply]
Isn't that because of WinUI3? Or is it strictly MAUI issue?

**Reply by liebki** [reply]
I understand, that is just a way to get the application to an user somehow.

> The thing the guy is getting at is there is currently no option for deploying to binary executable on the target platform as a stand alone app. You have to publish it to app stores in every platform to get the best experience or you can let people install it as an 'app' that's supposed to be more modern than the good old '.exe' but it requries you to have the certificate for this modern experience.
> 
> 
> 
> Can we just not have this modern experience though

**Reply by Xyncgas** [reply]
Thanks for your article

**Reply by Xyncgas** [reply]
Though why should one ever pay for a signed certificate, it's as simple as deriving from the root certificate that would be trusted on Microsoft's PC that they have established a chain of trust with the preinstalled certificates on windows, almost as if the programmer has a duty to be the visionary men Paul. G. Allen and Bill Gates once were in their little room and break free from this and make a distributed certificate system

**Reply by Xyncgas** [reply]
Imagine everyone can verify if a software is malicious by
1.Associating trusted parties, the parties should ensure these software they sign are not malicious or if so revoke their status or take actions
2.You can assume there aren't trusted parties, maybe some you trust so you verify if the software is signed by them, but most importantly establish multiple chain of trust yourself as opposed to the current landscape of preinstalled certificate on windows and the chain of trust being established by a few root certificate issuer this way
3.So everyone build their impression on what entity to trust, instead of centralized trusts through root certificate distribution on Windows, and allow some mechanism to convey these trusts between members of a human community & initiate this trust by the developer to the customers so if it's their first encounter of this software they are going to get scammed but let it be up for the community to gradually develop their policies to say not to install any software aside from their own trusted sources' referrals and share with other peers if a party has a bad reputation etc (it's not perfect)

**Reply by admoss0** [reply]
With other programming options I don't have to pay money to let others run my software. With .NET Maui I do.
So I choose other options.

My current strategy is to write a Blazor application rather than a Maui application. Blazor can do server based apps or web assembly apps that can run offline. Same C# ecosystem as Maui but no need to sign the code.

Blazor applications can be turned into Maui applications with few changes. But why would I do that if I have to pay to let other people run my programs?

**Reply by liebki** [reply]
> With other programming options I don't have to pay money to let others run my software. With .NET Maui I do.
> 
> So I choose other options.
> 
> 
> 
> My current strategy is to write a Blazor application rather than a Maui application. Blazor can do server based apps or web assembly apps that can run offline. Same C# ecosystem as Maui but no need to sign the code.
> 
> 
> 
> Blazor applications can be turned into Maui applications with few changes. But why would I do that if I have to pay to let other people run my programs?
> 
> 
> 
> 

So you basically mean web applications..? Because MAUI is native and ASP/Blazor is web, of course you don't need to pay for a certificate etc. here (if you don't sell and distribute the assemblies).

Signing of executables is a normal thing, what MAUI "does" is just a normal thing too, those certificates are just the normal way that's nothing new..

But I think the use of ASP.NET is a good way

**Reply by Xyncgas** [reply]
You can distribute MAUI by using dotnet publish, it outputs to publish directory with your executable and its dll, you can just copy that to people and they can just run it.

In the long run, if you need access to nativeAPI browser may or may not be enough, so ideally you would run your program in .NET running in Multi-threaded Webassembly in native wasm enviornment which provides fine grained access control therefore you don't need certificate to make a program safe

you are not paying microsoft to publish your program, if you require security from certificate signing, you can always do it yourself

**Reply by liebki** [reply]
> You can distribute MAUI by using dotnet publish, it outputs to publish directory with your executable and its dll, you can just copy that to people and they can just run it.
> 
> 
> 
> In the long run, if you need access to nativeAPI browser may or may not be enough, so ideally you would run your program in .NET running in Multi-threaded Webassembly in native wasm enviornment which provides fine grained access control therefore you don't need certificate to make a program safe
> 
> 
> 
> you are not paying microsoft to publish your program, if you require security from certificate signing, you can always do it yourself

Yea that's why I wrote that article, to explain how easy the deployment of MAUI is (without certificates). This doesn't mean that certificates are useless tho..

**Reply by Xyncgas** [reply]
certificate is useful, you're free to sign a certificate of anything you want, paying microsoft has the advantage of being trusted by the pre-installed root certificates on windows of course but you don't need those

**Comment by KennethHoff** [comment]
Before it supports Linux I won't even consider it.

**Reply by Hooterr** [reply]
There you go. I use MAUI purely for mobile, and I don't care about linux, macos, windows, tizen, web asm or any of that. Android and iOS are the only thing that matters to me.
I feel like MAUI is trying so hard to be one solution for everything. But with so many platforms supported it'd be a HUGE challenge for quite a big team but with the resources they have I don't know how do they plan to pull that off? Well... I know how... with questionable quality,

**Reply by developer9969** [reply]
Spot on!!! 
Grid layout is broken
Keyboard overlapping not fixed
Border clipping images and text
Binding releases issues 
Collectionview speechless 
Etc...
The above is net 7

Net 8
You can't even build the blimming app using a pc 

And all we see is silly videos on how todo the most trivial things

And then I a couple weeks we hear the same mantra "preview 7" and how wonderful that is .just hope they test it before shipping

**Reply by KennethHoff** [reply]
MAUI was always a highly questionable focus for the AspNetCore team, and they've clearly realized it.

**Reply by Stedy59** [reply]
??? Broad brush stroke there!

**Reply by Xyncgas** [reply]
Is it possible it's those situations when people don't know how to fix things, but instead of figuring out, the focuses are now on saying everything is good, maybe due to the people in the house feel like this is the best thing to do, which doesn't convey well to the customers.

Remember how CMS & CGI Federal told people healthcare.gov was going to work but it was messy, sometimes the problem was problem not being solved and that required a greater effort to address developer's problem instead of expecting them to solve the problems, I suspect the MAUI team can't help themselves and feel trapped and they need something we can give which we can if someone just come out today say it's not ready (doesn't have to say you made a mistake or you messed up, we know but it's ok) or cross platform UI is a bad idea or our approach is wrong we are ditching XAML in favors of html / we combine C# with XAML like C# with html in blazor / we are going to write cross platform UI completely in code and no more components to maintain therefore no more bugs for the teams to fix for components not working and instead just : {Provide a set of API for people to draw rectangle on screen, detect input event, access other features such as file system or camera or finger print etc if possible}

**Comment by nextcodelab** [comment]
From C# maui to dart Flutter.. all designs in flutter without lot of efforts is posible,,, now developing in flutter is maybe for me is 50x faster, plus the help of AI as code generator.
[https://play.google.com/store/apps/details?id=com.nxtccode.biblenotebook](https://play.google.com/store/apps/details?id=com.nxtccode.biblenotebook)

**Reply by liebki** [reply]
It seems like you're just trying to advertise your own application. Your application is not a good comparison to big company applications, you just can't transition from one language to another seamlessly, that would just interrupt the whole flow.

**Reply by nextcodelab** [reply]
I am not advertising it; I am using it for my own purposes. I just wanted to show that Flutter now works well for me and that the startup time is faster. I don't know if Maui will ever match Flutter's performance. I have already posted a video comparing the startup times of Flutter and Maui. In the comparison, even in release mode, the calculator app takes only one second to start in Flutter, whereas in Maui, it takes three to sometimes five seconds. Thank you."
[Flutter vs Maui start up in release mode](https://github.com/dotnet/maui/discussions/13043#discussioncomment-5344355)

**Reply by liebki** [reply]
You compare two different programming language is it from frameworks. 

I get that you're attracted to flutter and you think it's good, you just have to see the other side is that C#
and Maui are a big ecosystem and you get more if you just stay inside .net

For many companies or developers, there is no reason to just use another framework, because even if Maui is not good, or great or production ready, you can use it, and the whole power of C# combined

(That is my point of view, Flutter may be faster or smaller but if you work with .net, flutter is just not .net)

**Reply by Hooterr** [reply]
That's exactly right. But I wonder if there is a point of diminishing returns when you spend more time fixing bugs or finding workarounds in the framework than actually doing your work. 

Also, different apps are more suitable for different technologies. If all you do is make http calls, and parse JSON, there's no benefit in using .NET ecosystem because those are such basic tasks that any framework can do it. In that case you benefit HUGELY by using Flutter because you have a solid UI framework that always works. 
If you had some complex code sharing between backend and mobile, or required some specific .NET libraries then sure, MAUI is the best option with all it's shortcomings on the UI side.

Language or framework is just a tool. Choose the best one for the job.

**Reply by liebki** [reply]
I get you, I just don't like the never ending "X is better than MAUI". Give it some time or if you truly need something perfect, use something else like Avalonia etc.

In the end it depends on the developers or the available skill set 🦉

**Reply by Xyncgas** [reply]
Well I am still using MAUI and my app works so please don't break it like windows phone, but if I see MAUI canceled tomorrow I can switch to another platform because I have my own renderer for UI that will render my application in exactly the same way on any platform as long as they let me draw a rectangle and notify me on some events

**Comment by admoss0** [comment]
I've been writing in C# for many years and never been forced to code sign to run outside my IDE. Until Maui.Compulsory code signing be a thing of the future for big commercially applications, but I will not be doing it for the useful free small applications I write.For example, decades ago I wrote Wikicite, a C# program the helped people properly reference in Wikipedia articles. People throughout the world used my exe, and others ported my code to Linux and Mac with Mono.  Mobile devices weren't a thing then. I think I had a Nokia flip phone. I would not have bothered writing it if I had to pay $550 per year so other people could use it.BTW I've tried that advice to build Maui from the .Net command line with options to prevent use of a framework. It compiles OK but will not run outside the IDE. If you zip up the publish folder and move it to another machine it lacks necessary files.So, no Maui for me. C# still works, for now. I haven't been so disappointed since Oracle screwed up Java.DM On 7 Aug 2023, at 12:04 am, Kim ***@***.***> wrote:﻿

With other programming options I don't have to pay money to let others run my software. With .NET Maui I do.
So I choose other options.
My current strategy is to write a Blazor application rather than a Maui application. Blazor can do server based apps or web assembly apps that can run offline. Same C# ecosystem as Maui but no need to sign the code.
Blazor applications can be turned into Maui applications with few changes. But why would I do that if I have to pay to let other people run my programs?

So you basically mean web applications..?
Signing of executables is a normal thing, what MAUI "does" is just a normal thing too, those certificates are just the normal way that's nothing new..
But I think the use of ASP.NET is a good way

**Comment by admoss0** [comment]
You can write a Blazor app as a server app, which is a one page website. You can also write it as a Blazor Web Assembly, which can run locally without the need for Web connectivity. DM On 7 Aug 2023, at 12:04 am, Kim ***@***.***> wrote:﻿

With other programming options I don't have to pay money to let others run my software. With .NET Maui I do.
So I choose other options.
My current strategy is to write a Blazor application rather than a Maui application. Blazor can do server based apps or web assembly apps that can run offline. Same C# ecosystem as Maui but no need to sign the code.
Blazor applications can be turned into Maui applications with few changes. But why would I do that if I have to pay to let other people run my programs?

So you basically mean web applications..?
Signing of executables is a normal thing, what MAUI "does" is just a normal thing too, those certificates are just the normal way that's nothing new..
But I think the use of ASP.NET is a good way

**Comment by Xyncgas** [comment]
MAUI had a bad wild fire just now, some say that's a sign...

**Comment by waseem852** [comment]
RTL Support Sux Bigtime. and no one even cares to address those issues.
